### ISSUE TITLE
What's the capability of the rust-e1000 driver for now?

### ISSUE THREAD

**Issue by Richardhongyu** [fujita/rust-e1000#5]
Hi! The rust-e1000 driver is a fantastic job! And I want to run it on a real machine and have some tests with the c version.

With some modifications, I have successfully insmod the rust-e1000 driver and seen the nic info with the `ifconfig` command. But I failed to use this driver to ping a public IP address. It turns out that only the number of TX packets grows while the number of RX packets remains zero all the time. I also test this driver in the qemu environment. The result is the same. I see the `clean` function implementation for the E1000 struct that is supposed to support receiving the packages. But the number of RX packets failed to grow up.

I wonder whether something is wrong with my operation or if the rust-e1000 driver cannot receive packages due to the lack of some functions. Thanks for your help!

**Comment by fujita** [comment]
Ping works with QEMU. There are many MAC and phy types in e1000 NIC. I implemented a way to control MAC and phy types of QEMU e1000 NIC. I suspect your hardware are different.

**Comment by Richardhongyu** [comment]
I bought a NIC that has the chip version of 82540em. Are there other factors I need to care about?

Also, I can not use ping in the qemu environment also. My qemu command is 
```qemu-system-x86_64 \
        -nographic \
        -smp 1 -m 4096 \
        -kernel arch/x86_64/boot/bzImage \
        -append "console=ttyS0 root=/dev/sda1" \
        -hda ../x64_kernel.img \
        -netdev user,id=mynet0,hostfwd=tcp::8022-:22,hostfwd=tcp::8090-:80 \
        -device virtio-net-pci,netdev=mynet0
        -nic tap,model=e1000
```
Is there something that I miss?

**Comment by fujita** [comment]
With the standard e1000 driver, pings works on the above QEMU environment? Better to make sure that the network works correctly.
FYI, my command line is
```
qemu-system-x86_64 -smp 2 -m 1G -kernel "arch/x86_64/boot/bzImage" -initrd ../../initrd.img -nographic -vga none -append console=ttyS0 -no-reboot -nic tap -virtfs local,path=/home/ubuntu/git/quinn,mount_tag=quinn,security_model=none
```

**Comment by fujita** [comment]
The above command line helps?

**Comment by Richardhongyu** [comment]
> The above command line helps?

Sorry for the delay. I found that I didn't configure the tap device right. When I'm done, I will let you know.

**Comment by Richardhongyu** [comment]
Thanks for your help! I successfully ping an address using rust-e1000 driver.  I didn't configure the tap device correctly. After fixing this, everything is working correctly.

**Comment by stappersg** [comment]
> After fixing tap device, everything is working correctly.

Please add to the [README.md](https://github.com/fujita/rust-e1000/#readme) how a working tap device should look and can be configured.

**Comment by Richardhongyu** [comment]
> Ping works with QEMU. There are many MAC and phy types in e1000 NIC. I implemented a way to control MAC and phy types of QEMU e1000 NIC. I suspect your hardware are different.

My NIC has a `mac_type` of `e1000_82540` and a `phy_type` of `e1000_phy_m88`. I can accept packages but they are all dropped. Do you have any other clues about this result?

**Comment by fujita** [comment]
Might to be a good idea to debug if interrupts are properly received.

**Comment by Richardhongyu** [comment]
> Thanks for your help! I successfully ping an address using rust-e1000 driver. I didn't configure the tap device correctly. After fixing this, everything is working correctly.

When I debug, I found that I made a mistake. I did not unset the e1000 driver of the c version in the menuconfig. So after insmoding the rust-e1000 driver, the kernel will still use the e1000 driver of the c version. 

After fixing this, with the same config, c-e1000 can ping but rust-e1000 failed. The problem is the same as the real nic. Is this related to the qemu version? My qemu version is 6.2.0.
